packet-v2.go

```diff
@@ -0,0 +1,90 @@
+package macaroon
+
+import (
+	"encoding/binary"
+	"fmt"
+)
+
+type fieldType int
+
+// Field constants as used in the binary encoding.
+const (
+	fieldEOS            fieldType = 0
+	fieldLocation       fieldType = 1
+	fieldIdentifier     fieldType = 2
+	fieldVerificationId fieldType = 4
+	fieldSignature      fieldType = 6
+)
+
+type packetV2 struct {
+	// fieldType holds the type of the field.
+	fieldType fieldType
+
+	// data holds the packet's data.
+	data []byte
+}
+
+// parseSectionV2 parses a sequence of packets
+// in data. The sequence is terminated by a packet
+// with a field type of fieldEOS.
+func parseSectionV2(data []byte) ([]byte, []packetV2, error) {
+	prevFieldType := fieldType(-1)
+	var packets []packetV2
+	for {
+		if len(data) == 0 {
+			return nil, nil, fmt.Errorf("section extends past end of buffer")
+		}
+		rest, p, err := parsePacketV2(data)
+		if err != nil {
+			return nil, nil, err
+		}
+		if p.fieldType == fieldEOS {
+			return rest, packets, nil
+		}
+		if p.fieldType <= prevFieldType {
+			return nil, nil, fmt.Errorf("fields out of order")
+		}
+		packets = append(packets, p)
+		prevFieldType = p.fieldType
+		data = rest
+	}
+}
+
+func parsePacketV2(data []byte) ([]byte, packetV2, error) {
+	data, ft, err := parseVarint(data)
```

Comment:
Not as far as I can see. The second value returned by parseVarint is the number we've read, not the number of read bytes.